\documentclass[11pt]{article}

% \usepackage[UTF8]{ctex} % for Chinese 

\usepackage{setspace}
\usepackage[colorlinks,linkcolor=blue,anchorcolor=red,citecolor=black]{hyperref}
\usepackage{lineno}
\usepackage{booktabs}
\usepackage{graphicx}
\usepackage{float}
\usepackage{floatrow}
\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{geometry}
\usepackage{multirow}
\usepackage{longtable}
\usepackage{lscape}
\usepackage{booktabs}
\usepackage{natbibspacing}
\usepackage[toc,page]{appendix}
\usepackage{makecell}
\usepackage{amsfonts}
 \usepackage{amsmath}
\usepackage[utf8]{inputenc}
\usepackage{amssymb}
\usepackage{amsthm}
\usepackage{enumerate}
\usepackage{comment}

\usepackage[backend=bibtex,style=authoryear,sorting=nyt,maxnames=1]{biblatex}
\bibliography{README} % Reference bib

\title{Pipelines}
\author{}
\date{}

\linespread{1.5}
\geometry{left=2cm,right=2cm,top=2cm,bottom=2cm}

\setlength\bibitemsep{0pt}

\begin{document}
\begin{sloppypar}
  \maketitle

  \linenumbers
\section{DeCon}
\subsection{Introduction}
DeCon is designed to retrieve genomic sequences of a target \textbf{phylum} from metagenomic assembly of paired next generation sequencing (NGS) reads. 
First, contigs below 400 base pair (bp) are removed and NGS reads are mapped to filtered assembly by minimap2 \parencite{li2018minimap2}. 
Second, SprayNPray \parencite{garber2022spraynpray} is used to compute coverage, GC content and coding density of each contigs. 
Third, contigs are searched against non-redundant (nr) database by DIAMOND \parencite{buchfink2015fast} in long read mode and assigned to phyla by MEGAN \parencite{huson2007megan} in long read mode. 
Forth, a decision tree classifier is trained, taking coverage, GC content and coding density of contigs as training features and phylum assignments from MEGAN as target value. 
This classifier is used to compute phylum assignments of contigs that are not determined by MEGAN. 
Fifth, contigs assigned to the target phylum are retrieved. 
QUAST \parencite{gurevich2013quast} and BUSCO \parencite{simao2015busco} are used to evaluate retrieved genome. 
\subsection{Dependencies}
\textbf{Softwares} \newline
R \newline
Python \newline
minimap2 \newline
SAMtools \newline
SprayNPray \newline
DIAMOND \newline
MEGAN tools (daa-meganizer & daa2info) \newline
seqkit \newline
QUAST \newline
BUSCO \newline
\par
\textbf{Databases} \newline
DIAMOND database (nr) \newline
MEGAN database \newline
BUSCO database \newline
\par
\textbf{Python modules} \newline
numpy \newline
pandas \newline
scikit-learn \newline
\par
\textbf{R packages} \newline
reticulate \newline
stringr \newline
\par
\subsection{Usage}
Modify configuration file (templated as DeCon\_conf.R), and run \newline
Rscript \textif{path}/DeCon\_pipeline.R \textif{path}/DeCon\_main.R \textif{path}/DeCon\_main.py \textif{path}/DeCon\_conf.R

\section{ProtGeneCall}
\subsection{Introduction}
ProtGeneCall is designed to call protein-coding genes from genome, combining protein-genome alignments, transcriptome-genome alignments and \textit{ab initio} gene predictions. 
First, repeat elements are identified by RepeatModeler \parencite{smit2015repeatmodeler} and masked by RepeatMasker \parencite{smit2015repeatmasker}. 
Masked genome is used for downstream analysis. 
Second, proteins of closely related species are mapped to the masked genome by miniprot \parencite{li2023protein}. 
Third, paired RNA-sequencing (RNA-seq) reads are mapped to masked genome by Hisat2 \parencite{kim2019graph}. 
Forth, transcriptome-genome alignments are computed by StringTie \parencite{pertea2015stringtie}. 
Fifth, gene structures are predicted from transcriptome-genome alignments by TransDecoder \parencite{haas2016transdecoder}, combining searching against UniRef \parencite{suzek2007uniref} by \parencite{buchfink2015fast} and PfamA \parencite{mistry2021pfam} by HMMER \parencite{eddy1992hmmer}. 
Sixth, AUGUSTUS \parencite{stanke2003gene} is trained with gene structures from TransDecoder to compute gene predictions. 
Seventh, BRAKER \parencite{hoff2019whole} is trained with RNA-seq mapping to call genes. 
Eighth, GALBA \parencite{hoff2019whole} is trained with proteins of closely related species. 
Ninth, protein-genome alignments from miniprot, transcript-genome alignments from Hisat2-StringTie, and \textit{ab initio} gene predictions from TransDecoder, AUGUSTUS, BRAKER and GALBA are integrated into consensus gene structures by EvidenceModeler \parencite{haas2008automated}. 
Tenth, genes from EvidenceModeler are removed if they are supported by only one \textit{ab initio} predictor and lack protein/RNA-seq evidence. 
Eleventh, two iterations of PASA \parencite{haas2008automated} is used to update filtered gene structures from EvidenceModeler. 
Twelfth, genes with in-frame stop codons or incomplete coding regions (coding regions with length cannot be divided by 3) are removed and the predicted peptide set is evaluated by BUSCO \parencite{simao2015busco}.
\subsection{Dependencies}
\textbf{Softwares} \newline
R \newline
Python \newline
RepeatModeler \newline
RepeatMasker \newline
miniprot \newline
Hisat2 \newline
SAMtools \newline
StringTie \newline
TransDecoder \newline
HMMER \newline
DIAMOND \newline
AGAT \newline
AUGUSTUS \newline
BLAST+ \newline
GALBA \newline
BRAKER \newline
EvidenceModeler \newline
BUSCO \newline
gffread \newline
seqkit \newline
MAKER \newline
\par
\textbf{Databases} \newline
DIAMOND database (UniRef) \newline
Pfam-A \newline
\par
\textbf{External scrpts} \newline
cufflinks\_gtf\_to\_alignment\_gff3.pl from EvidenceModeler \newline
augustus\_GFF3\_to\_EVM\_GFF3.pl from EvidenceModeler \newline
gth2gtf.pl from AUGUSTUS \newline
computeFlankingRegion.pl from AUGUSTUS \newline
gff2gbSmallDNA.pl from AUGUSTUS \newline
gtf2aa.pl from AUGUSTUS \newline
simplifyFastaHeaders.pl from AUGUSTUS \newline
aa2nonred.pl from AUGUSTUS \newline
filterGenesIn.pl from AUGUSTUS \newline
autoAug.pl from AUGUSTUS \newline
evm\_evidence.py in this GitHub \newline
\par
\textbf{R packages} \newline
stringr \newline
parallel \newline
\par
\subsection{Usage}
Modify configuration file (templated as ProtGeneCall\_conf.R), and run \newline
Rscript \textif{path}/ProtGeneCall\_pipeline.R \textif{path}/ProtGeneCall\_main.R \textif{path}/ProtGeneCall\_conf.R

\section{Pipeline for calling repeat elements from genome (RepCall)}
\subsection{Introduction}
Pipeline RepCall is designed to call repeat elements genes from genome. 
First, miniature inverted-repeat transposable elements (MITE) are called by MITE-Hunter \parencite{han2010mite}. 
Second, long terminal repeats (LTRs) are identified by incorporating LTR\_FINDER\_parallel \parencite{ou2019ltr_finder_parallel}, LTRharvest \parencite{ellinghaus2008ltrharvest} and LTR\_retriever \parencite{ou2018ltr_retriever}. 
Third, identified MITEs and LTRs are masked by RepeatMasker \parencite{smit2015repeatmasker}. 
Forth, RepeatModeler \parencite{smit2015repeatmodeler} is used to further identify repeats in the masked genome. 
Fifth, the locations of MITEs, LTRs and repeats from RepeatModeler are identified by RepeatMasker and all repeats are incorporated into a consensus library. 
\subsection{Dependencies}
\textbf{Softwares} \newline
R \newline
seqkit \newline
MITE-Hunter \newline
LTR\_FINDER\_parallel \newline
LTRharvest \newline
LTR\_retriever \newline
RepeatMasker \newline
RepeatModeler \newline
\par
\subsection{Usage}
Modify configuration file (templated as RepCall\_conf.R), and run \newline
Rscript \textif{path}/RepCall\_pipeline.R \textif{path}/RepCall\_main.R \textif{path}/RepCall\_conf.R

\section{ncRNAcall}
\subsection{Introduction}
ncRNAcall is designed to call non-coding RNA (ncRNA) from genome. 
First, transfer RNA (tRNA) is identified by tRNAscan-SE \parencite{lowe1997trnascan}. 
Second, microRNA is called by miRNAture \parencite{velandia2021mirnature}. 
Third, target genes of microRNA are identified by searching microRNA against annotated three prime untranslated regions (3'UTR) by miRanda \parencite{enright2003microrna}. 
Forth, Infernal \parencite{nawrocki2013infernal} searches against Rfam \parencite{kalvari2021rfam} database to call other non-coding RNA, \textit{e.g.} ribosomal RNA (rRNA) and small nuclear RNA (snRNA). 
Fifth, all results are incorporated together. 
\subsection{Dependencies}
\textbf{Softwares} \newline
R \newline
tRNAscan-SE \newline
biocode \newline
miRNAture \newline
miRanda \newline
bedtools \newline
seqkit \newline
Infernal \newline
\par
\textbf{Databases} \newline
miRNAture database \newline
Rfam database \newline
\par
\textbf{R packages}
parallel \newline
stringr \newline
\par
\subsection{Usage}
Modify configuration file (templated as ncRNAcall\_conf.R), and run \newline
Rscript \textif{path}/ncRNAcall\_pipeline.R \textif{path}/ncRNAcall\_main.R \textif{path}/ncRNAcall\_conf.R

\section{buscoProt2Phylo}
\subsection{Introduction}
buscoProt2Phylo infers phylogenetic tree using single-copy genes defined by BUSCO \parencite{simao2015busco}. 
First, from BUSCO runs complete single-copy protein sequences are collected and classified according to protein families that they belong to. 
Second, for protein famiies that are identified in above 4 BUSCO runs, protein sequences are aligned by MAFFT \parencite{katoh2002mafft}. 
Third, multiple sequence alignments from MAFFT are trimmed by trimAl \parencite{capella2009trimal}. 
Forth, gene trees are inferred from trimmed multiple sequence alignments by IQ-TREE \parencite{minh2020iq} with 1,000 bootstrap replicates. 
Fifth, species tree is inferred from gene trees by ASTRAL \parencite{zhang2018astral}. 
Sixth, a supermatrix method was used to infer species tree from the multiple sequence alignments from MAFFT. 
Multiple sequence alignments contains 85\%, 87.5\%, 90\% 92.5\%, 95\%, 97.5\% and 100\% of the total species were concatenated into supermatrixes, respectively. 
Missing species were represented by gaps. 
From each supermatrix a species tree was inferred by IQ-TREE \parencite{minh2020iq} with 1,000 bootstrap replicates. 
\subsection{Dependencies}
\textbf{Softwares} \newline
R \newline
MAFFT \newline
trimAl \newline
seqkit \newline
IQ-TREE \newline
ASTRAL \newline
\par
\textbf{R packages} \newline
parallel \newline
\par
\subsection{Usage}
Modify configuration file (templated as buscoProt2Phylo\_conf.R), and run \newline
Rscript \textif{path}/buscoProt2Phylo\_pipeline.R \textif{path}/nbuscoProt2Phylo\_main.R \textif{path}/buscoProt2Phylo\_conf.R

\section{metaTrans}
\subsection{Introduction}
metaTrans is designed for taxonomic profiling of metatranscriptomic sequencing of paired NGS reads. 
First, metatranscriptomic reads are mapped to corresponding host genome by Hisat2 \parencite{kim2019graph} and unmapped reads are extracted by SAMtools \parencite{li2009sequence}. 
Second, ribosomal RNA reads are removed by SortMeRNA \parencite{kopylova2012sortmerna}. 
Third, all reads are pooled together and assembled by rnaSPAdes \parencite{bushmanova2019rnaspades}. 
Forth, MMseqs2 \parencite{steinegger2017mmseqs2} (--cov-mode 1 -c 0.75 --min-seq-id 0.75) is used to remove redundancy in assembly from rnaSPAdes. 
Fifth, coding regions of assembled transcripts are identified by TransDecoder \parencite{haas2016transdecoder}, combining searching against UniRef \parencite{suzek2007uniref} by DIAMOND \parencite{buchfink2015fast} and PfamA \parencite{mistry2021pfam} by HMMER \parencite{eddy1992hmmer}. 
Sixth, protein sequences transcribed by assembled transcripts are searched against non-redundant database by DIAMOND \parencite{buchfink2015fast} and assigned to taxa by MEGAN \parencite{huson2007megan}. 
Seventh, reads are mapped to assembled transcripts by minimap2 \parencite{li2018minimap2} and SAMtools is used to compute coverage and depth of transcripts. 
Eighth, coverage, depth, coordinates of coding regions and taxonomy assignments of transcripts are taken together as comprehensive tables. 
Ninth, protein functions are inferred by InterproScan \parencite{jones2014interproscan}. 
Tenth, protein functions are inferred by eggNOG-mapper \parencite{cantalapiedra2021eggnog}. 
\subsection{Dependencies}
\textbf{Softwares} \newline
R \newline
Hisat2 \newline
SAMtools \newline
SortMeRNA \newline
SPAdes \newline
MMseqs2 \newline
TransDecoder \newline
DIAMOND \newline
HMMER \newline
MEGAN tools (daa-meganizer & daa2info) \newline
minimap2 \newline
gffread \newline
seqkit \newline
MAKER \newline
InterproScan \newline
eggNOG-mapper \newline
\par
\textbf{External scripts}
simplifyFastaHeaders.pl from AUGUSTUS \newline
\par
\textbf{Databases} \newline
DIAMOND database (UniRef) \newline
Pfam-A \newline
\par
\textbf{R packages} \newline
stringr \newline
\subsection{Usage}
Modify configuration file (templated as metaTrans\_conf.R), and run \newline
Rscript \textif{path}/metaTrans\_pipeline.R \textif{path}/metaTrans\_main.R \textif{path}/metaTrans\_conf.R

\section{PseudoCall}
\subsection{Introduction}
PseudoCall is designed to call pseudogenes with PseudoPipe that has been modified to (1) use stricter criteria for filtering blast hits, (2) run commands in parallel and (3) enable restarting. (\textbf{To be continued...})
\subsection{Dependencies}
\subsection{Usage}
Modify configuration file (templated as PseudoCall\_conf.R), and run \newline
Rscript \textif{path}/PseudoCall\_pipeline.R \textif{path}/PseudoCall\_main.R \textif{path}/PseudoCall\_conf.R

% Modify configuration file (templated as ncRNAcall\_conf.R), and run \newline
% Rscript \textif{path}/ncRNAcall\_pipeline.R \textif{path}/ncRNAcall\_main.R \textif{path}/ncRNAcall\_conf.R

% \section{ncRNAcall}
% \subsection{Introduction}
% \subsection{Dependencies}
% \subsection{Usage}
% Modify configuration file (templated as ncRNAcall\_conf.R), and run \newline
% Rscript \textif{path}/ncRNAcall\_pipeline.R \textif{path}/ncRNAcall\_main.R \textif{path}/ncRNAcall\_conf.R

\printbibliography
\end{sloppypar}
\end{document}