\documentclass[11pt]{article}

% \usepackage[UTF8]{ctex} % for Chinese 

\usepackage{setspace}
\usepackage[colorlinks,linkcolor=blue,anchorcolor=red,citecolor=black]{hyperref}
\usepackage{lineno}
\usepackage{booktabs}
\usepackage{graphicx}
\usepackage{float}
\usepackage{floatrow}
\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{geometry}
\usepackage{multirow}
\usepackage{longtable}
\usepackage{lscape}
\usepackage{booktabs}
\usepackage{natbibspacing}
\usepackage[toc,page]{appendix}
\usepackage{makecell}
\usepackage{amsfonts}
 \usepackage{amsmath}

\usepackage[backend=bibtex,style=authoryear,sorting=nyt,maxnames=1]{biblatex}
\bibliography{} % Reference bib

\title{Abstracts: Hemiptera}
\author{}
\date{}

\linespread{1.5}
\geometry{left=2cm,right=2cm,top=2cm,bottom=2cm}

\setlength\bibitemsep{0pt}

\begin{document}
\begin{sloppypar}
  \maketitle

  \linenumbers
\textbf{Altincicek \textit{et al.}, 2008, Wounding-mediated gene expression and accelerated viviparous reproduction of the pea aphid Acyrthosiphon pisum.} \newline
Piercing of the pea aphid \textit{Acyrthosiphon pisum} with a bacteria-contaminated needle elicits lysozyme-like activity in the haemolymph but no detectable activities against live bacteria. 
No homologues of known antimicrobial peptides were found in cDNA library generated by using the suppression subtractive hybridization method or in over 90 000 public expressed sequence tag (EST) sequences, but lysozyme genes have recently been described in pea aphid. 
Production of viviparous offspring was significantly accelerated upon wounding.
\par
Pea aphid showed weakened immune system. 
No homologues of known antimicrobial peptides were found. 
The presence of insect defensins in other Hemiptera and in the basal apterygote insect \textit{Thermobia domestica} (Altincicek & Vilcinskas, 2007) suggests that at least this type of antimicrobial peptides may have been lost during aphid evolution. 
Interestingly, the observation that pierced aphids showed a limited capacity to seal their wound by haemolymph coagulation and melanization agrees with the finding that an encapsulation response of pea aphid to the parasitoid wasp \textit{Aphidius ervi} is either very weak or non-existent (Oliver et al., 2005).
\par
Regards to weakened immunity of pea aphids: 
(1) Aphids and relatives of Hemiptera share the unique ability to exploit exclusively phloem sap as diet, which is usually sterile (Douglas, 2006). Thus, the risk of encountering pathogens in their diet is limited. 
(2) Aphids harbour primary symbionts that are vertically transmitted and located intracellularly, as well as secondary symbionts that are both vertically and horizontally transmitted and also survive extracellularly in the insect haemolymph where they face the host's antimicrobial defences (Moran & Dunbar, 2006; Haine, 2008). 
It is possible that the symbionts provide protection, \textit{e.g.} pea aphid has been reported to be protected against fungal pathogens by the facultative symbiotic Gram-negative bacterium \textit{Regiella insecticola} (Scarborough et al., 2005) and also against the parasitoid wasp \textit{Aphidius ervi} by the facultative symbiotic Gram-negative bacterium \textit{Hamiltonella defensa} (Oliver et al., 2005). 
This may further explain why only lysozyme-like activity is present in the haemolymph, as lysozymes target mainly Gram-positive bacteria, whereas aphid symbionts belong to Gram-negative bacteria. 
(3) As immune responses are costly because they require investment of resources which are shared with other fitness-relevant traits (Rolff & Siva-Jothy, 2003; Schmidt-Hempel, 2005; Freitak et al., 2007), it is reasonable that aphids increase terminal reproductive investment in response to a putative survival threat such as an immune challenge. 
\par
\textbf{Gerardo \textit{et al.}, 2010, Immunity and other defenses in pea aphids, Acyrthosiphon pisum.} \newline
Pea aphids appear to be missing genes present in insect genomes characterized to date and thought critical for recognition, signaling and killing of microbes. 
In line with results of gene annotation, experimental analyses designed to characterize immune response through the isolation of RNA transcripts and proteins from immune-challenged pea aphids uncovered few immune-related products. 
Gene expression studies, however, indicated some expression of immune and stress-related genes.
\par
In the fruit fly \textit{Drosophila melanogaster}, recognition of an invasive microbe leads to signal production via four pathways (Toll, IMD, JNK, and JAK/STAT) (Boutros \textit{et al.}, 2002). 
Each pathway is activated in response to particular pathogens (Dionne \textit{et al.}, 2008). 
Signaling triggers the production of multitude effectors, including, most notably, antimicrobial peptides (AMPs). 
In insect genomes annotated to date, these pathways appear well conserved, with most of the key components found across flies (\textit{Drosophila spp.}) (Sackton \textit{et al.}, 2007), mosquitoes (\textit{Aedes aegypti}, \textit{Anopheles gambiae}) (Waterhouse \textit{et al.}, 2007; Christophides \textit{et al.}, 2002), bees (\textit{Apis mellifera}) (Evans \textit{et al.}, 2006) and beetles (\textit{Tribolium castaneum}) (Zou \textit{et al.}, 2007).
\par
The cellular component of pea aphids’ innate immune response may also be different to that seen in other insects. 
While many insects encapsulate parasitoid wasp larvae, smothering them to death with hemocytes, aphids appear not to have this layer of protection (Bensadia \textit{et al.}, 2006; Carver \textit{et al.}, 1988). 
Aphids, however, appear to recruit some hemocytes to parasitoid eggs, suggesting that cellular immunity may play an alternative, though possibly more limited, role (Bensadia \textit{et al.}, 2006).
\par
There is evidence that pea aphid has some defense systems common to other arthropods, \textit{e.g.}, the Toll and JAK/STAT signaling pathways, HSPs, ProPO. 
However, several of the genes thought central to arthropod innate immunity are missing in pea aphid, including PGRPs, the IMD signaling pathway, defensins, c-type lysozymes. 
\par
The failure of finding aphid homologs to many insect immune genes can be resulted from large evolutionary distance between pea aphid and taxa used as reference (divided ~100 million years ago). 
However, similar homology-search based method successfully detected immune-related genes in even more divergent insects. 
Another explanation for lack of immune genes is that pea aphid mount an alternative but equal immunity. 
However, functional analysis, together with Altincicek \textit{et al.}, 2008, found little evidence for an alternative response to \textit{E. coli} infection. 
\par
Altincicek \textit{et al.}, 2008 proposed three hypotheses on the ecological success of pea aphid with the possibility of lacking a strong immunity. 
First, aphids feeds on plant sap which is often sterile, leading to reduced risk for encountering pathogens. 
However, aphids are capable of acquiring pathogenic bacteria from the surface of their host plants’ leaves (Stavrinides \textit{et al.}, 2009), and aphids become host to a diverse assemblage of bacteria and fungi under stressful conditions (Nakabachi \textit{et al.}, 2003). 
Furthermore, \textit{Sitophilus} weevils, which when challenged with \textit{E. coli} significantly up-regulate immune genes (Anselme \textit{et al.}, 2008), spend their entire larval and nymph stages within sterile cereal grains, indicating that a sterile diet is not likely to explain the absence of antibacterial defenses in aphids. 
Second, aphid symbionts may provide protection against pathogens, \textit{e.g.} pea aphid has been reported to be protected against fungal pathogens by the facultative symbiotic Gram-negative bacterium \textit{Regiella insecticola} (Scarborough et al., 2005) and also against the parasitoid wasp \textit{Aphidius ervi} by the facultative symbiotic Gram-negative bacterium \textit{Hamiltonella defensa} (Oliver et al., 2005). 
This seems plausible regards to the cost of immune gene expression versus the benefit of protection by the secondary endosymbionts. 
However, it does not explain how the secondary endosymbionts (as Gram-negative bacteria), often present in aphid hemolymph, are themselves perceived and controlled by the aphid immune system.
Third, aphids may invest in terminal reproduction in response to an immune challenge, rather than in a costly immune response, as Altincicek \textit{et al.}, 2008 found increased viviparous offspring production upon wounding. 
Such an increase has been found in many invertebrates including \textit{Biomphalaria} snails (Minchella \textit{et al.}, 1981; Minchella \textit{et al.}, 1985), \textit{Acheta} crickets (Adamo \textit{et al.}, 1999), \textit{Daphnia} waterfleas (Chadwick \textit{et al.}, 2005), and \textit{Drosophila} flies (Polak \textit{et al.}, 1998). 
Even without immune challenge, these insects also tends to invest most resources towards rapid, early onset reproduction (r-selection), and such organisms may specifically invest less in costly immune responses (Zuk \textit{et al.}, 2002; Miller \textit{et al.}, 2007). 
However, this may not sufficient for explaining weak immunity of pea aphids, as r-selected taxa such as \textit{Drosophila} still mount complex immune responses. 
Furthermore, aphids do not increase their reproductive effort in the face of all immune challenges: 
fungal infection reduces the number of offspring pea aphid produce within 24 hours of inoculation (Baverstock \textit{et al.}, 2006), and response to stabbing with bacteria seems to be specific to the aphid genotype and to the location of the stab.
\par
\textbf{Kim and Lee, 2017, Insect symbiosis and immunity: the bean bug-Burkholderia interaction as a case study.} \newline
Primary symbionts refers to maternally transmitted obligated symbionts. 
They associate with host at evolutionary time scale and become vital to host survival. 
The adaptive evolution of primary symbionts is accompanied with dramatic reduction in genome size and loss of genes essential for free-living (Moran, 2003; McCutcheon and Moran, 2012). 
Secondary or facultative symbionts are vertically/horizontally transmitted, and are not essential for host survival. 
Their association with hosts is recent, and still possess free-living ability without collapsed genomes.
\par
Bean bug \textit{Riptortus pedestris} is a member of order Hemiptera feeding on plant sap. 
The midgut of bean bug is divided into morphologically distinct regions called M1, M2, M3, M4B and M4, where M4 is symbiotic. 
Symbiotic organ in M4 has two rows of crypts, whose lumens are densely colonized by betaproteobacterial symbionts of genus \textit{Burkholderia}. 
The symbionts of bean bug are acquired orally from rhizosphere environment during early nymphal stage. 
\textit{Burkholderia} is a soil bacterium. 
They retain free-living ability after association with bean bug and are easily cultured in labs.
\par
nder bacterial challenges, symbiotic bean bugs exhibits better survival than aposymbiotic (Kim \textit{et al.}, 2015). 
This better survival retains after inhibition of cellular immunity, indicating stronger humoral immunity induced by \textit{Burkholderia}. 
Without bacterial challenges, antimicrobial peptide (AMP) (riptocin, rip-defensin and rip-thanatin) expression in fat body of symbiotic and aposymbiotic bean bugs is similar. 
However, AMP expression significantly increases in symbiotic bean bugs compared with aposymbiotic ones (Kim \textit{et al.}, 2015).
\par
\textit{Burkholderia} is Gram-negative bacterium. 
Its cell envelop consists of inner- and outer membranes. 
Lipopolysaccharide (LPS) is located at the outer part of outer membrane. 
It is composed of lipid A embedded in outer membrane and oligosaccharide connecting with O-antigen. 
In symbiotic \textit{Burkholderia}, O-antigen is lost when compared with free-living \textit {Burkholderia} (Kim \textit{et al.}, 2005). 
However, lipid A and oligosaccharide are retained. 
Besides, symbiotic \textit{Burkholderia} are more susceptible to detergent than free-living ones, indicating compromised cell membrane integrity.
\par
Free-living \textit{Burkholderia} are resistant to antimicrobial activity of bean bug haemolymph (Loutet and Valvano, 2011), while symbiotic ones are highly susceptible (Kim \textit{et al.}, 2015). 
When bean bug AMPs are purified, symbiotic \textit{Burkholderia} are more susceptible to riptoctin and rip-defensin than free-living ones (Kim \textit{et al.}, 2015). 
Ultimately, injected symbiont \textit{Burkholderia} are removed much faster by bean bug immunity than free-living ones (Kim \textit{et al.}, 2015).
\par
Systemic injection of symbiotic \textit{Burkholderia}, free-living \textit{Burkholderia} and \textit{Escherichia coli} triggers similar level of AMP expression in fat body. 
However, in midgut M4 region, the expression of AMPs is similar in symbiotic and aposymbiotic bean bugs. 
The AMP expression in M4 region is lower than basal expression of AMPs in fat body (Kim \textit{et al.}, 2015). 
These indicate potential immune privilege of symbiotic organ. 
\par
The susceptibility of \textit{Burkholderia} to bean bug immunity could be an advantage for easy management of symbiotic population. 
During nymphal stage, the size of symbiont population increases. 
However, pattern of transient decrease of symbiotic population is observed prior to moulting period in each instar stage. 
This transient decrease is corresponding to increase of antimicrobial activity of symbiont organ, including up-regulated expression of c-type lysosome and riptocin (Kim \textit{et al.}, 2014). 
Another mechanism for symbiont population management in bean bugs is related to M4B region of midgut. 
M4B region of symbiotic bean bug exhibits strong antimicrobial activity, while in aposymbiotic bean bug, little antimicrobial activity is exhibited. 
Besides, died bacterial symbionts present in M4B region. 
These indicate that the antimicrobial activity of M4B midgut is induced by \textit{Burkholderia} (Kim \textit{et al.}, 2013). 
One of components responsible for antimicrobial activity of M4B region is cathepsin-L-like protease (Byeon \textit{et al.}, 2015), which is highly and preferentially expressed in M4B region of symbiotic bean bugs (Futahashi \textit{et al.}, 2013). 
Antimicrobial activity exhibited by M4B and M4 region of bean bugs is only effective for symbiont \textit{Burkholderia}, but free-living \textit{Burkholderia} are resistant (Byeon \textit{et al.}, 2015; Kim \textit{et al.}, 2013). 
\par
\textbf{Husink and McCutcheon, 2016, Repeated replacement of an intrabacterial symbiont in the tripartite nested mealybug symbiosis.} \newline
Citrus mealybug \textit{Planococcus citri} has two bacterial endosymbionts with an unusual nested arrangement: the $\gamma$-proteobacterium \textit{Moranella endobia} lives in the cytoplasm of the $\beta$-proteobacterium \textit{Tremblaya princeps}. 
To test the stability of this three-way symbiosis, host and symbiont genomes for five diverse mealybug species were sequenced. 
$\beta$-proteobacterial genomes from diverse mealybug species are \textit{Tremblaya} with similar genome sizes, while $\gamma$-proteobacteria are from different clades with different genome sizes. 
Therefore, it is inferred that \textit{Tremblaya} is the result of a single infection in the ancestor of mealybugs, while the $\gamma$-proteobacterial symbionts result from multiple replacements of inferred different ages from related but distinct bacterial lineages. 
\par
Three scenario of the order and timing of the $\gamma$-proteobacterial infections are proposed. 
In idosyncratic scenario, there was a single $\gamma$-proteobacterial acquisition in the ancestor of the Pseudococcinae that has evolved idiosyncratically as mealybugs diversified over time, leading to seemingly unrelated genome structures and coding capacities. 
In independent scenario, the $\gamma$-proteobacterial infections occurred independently, each establishing symbioses inside \textit{Tremblaya} in completely unrelated and separate events. 
In replacement scenario, there was a single $\gamma$-proteobacterial acquisition in the Pseudococcinae ancestor that has been replaced in some mealybug lineages over time. 
\par
The idosyncratic scenario can be discarded as phylogenetics of $\gamma$-proteobacterial symbionts reveals that they have originated from clearly distinct and well-supported bacterial lineages. 
The independent and replacement scenarios are more difficult to tell apart. 
Under the independent scenario, \textit{Tremblaya} may experience two rounds of genome corruption: one after association with mealybug ancestor, and one after infection of $\gamma$-proteobacteria. 
Therefore, one should expect diverse genome sizes in $\beta$- and $\gamma$-proteobacteria. 
Conserved genome sizes of $\beta$-proteobacteria and diverse sizes of $\gamma$-proteobacterial genomes favor the replacement scenario. 
\par
Two reasons why $\gamma$-proteobacteria end with living inside $\beta$-proteobacteria are proposed. 
The first is that it was easier to use the established transport system between the insect cell and \textit{Tremblaya} than to evolve a new one. 
The second is that the insect immune system likely does not target \textit{Tremblaya} cells, and so the \textit{Tremblaya} cytoplasm is an ideal hiding place for a newly arrived symbiont.
\par
\textbf{Gil \textit{et al.}, 2017, Tremblaya phenacola PPER: an evolutionary beta-gamma-proteobacterium collage.} \newline
Bougainvillea mealybug \textit{Phenacoccus peruvianus} (PPER) harbors single betaproteobacterial symbiont \textit{Tremblaya phenacola}. 
The genome of \textit{Tremblaya phenacola} PPER is highly rearranged, in contrast to the high genomic stability of all previously sequenced \textit{Tremblaya} lineages, with an almost absolute synteny conservation among \textit{Tremblaya phenacola} strains (McCutcheon and von Dohlen, 2011; Husnik and McCutcheon, 2016), and a single inversion in \textit{Tremblaya phenacola} in \textit{Phenacoccus avenae} (PAVE) (Husnik and McCutcheon, 2016). 
Chromosome rearrangements cause perturbations in GC skew, which have a deleterious impact upon the replication system (Rocha, 2004). 
Therefore, although bacterial chromosomes can undergo many rearrangements at the beginning of an endosymbiotic relationship (see the marked example of ‘\textit{Candidatus} Sodalis pierantonius’ SOPE; Oakeson et al., 2014), long-term endosymbionts tend to present a typical GC skew, an indication that it is recovered with evolutionary time. 
Contrary to the PAVE genome, with a typical GC-skew pattern (Rocha, 2008), PPER genome presents a non-polarized and highly disrupted GC skew, except in the most syntenic region between both genomes, containing most ribosomal protein genes, suggesting that the chimeric genomic architecture is not stabilized.
\par
The \textit{Tremblaya phenacola} PPER genome contains 192 different CDSs, 188 with an assigned function. 
There are only four duplicated genes inside repeats (\textit{rpsU}, \textit{hisG}, \textit{prmC} and TPPER_00169/220), and two
have two homologs (\textit{infA} and \textit{rlmE}). 
It possesses a single ribosomal operon and a complete set of tRNA genes for all 20 amino acids, 4 of them inside repeats. 
\par
In annotated CDSs, 102 CDSs appear to be of betaproteobacterial origin, but another 80 appear to belong to a gammaproteobacterium. 
Furthermore, there is a relationship between the taxonomic affiliation of each identified CDS and their G+C content. 
Generally, genes with gammaproteobacterial assignation have lower G+C values than betaproteobacterial assignation (Agashe and Shankar, 2014), which is consistent with genes in \textit{Tremblaya phenacola} PPER. 
\textit{Tremblaya phenacola} PPER genes not assigned to any category have a wide range in G+C content, and most
of them have very short length. 
There is also differences in codon usage depending on the beta or gammaproteobacterial assignation in genes of \textit{Tremblaya phenacola} PPER genes. 
The distribution of gamma or beta genes along the \textit{Tremblaya phenacola} PPER genome is not random: most contigs contain only genes of one taxonomic origin, some others change the gene affiliation in the middle, and only one contig 
is completely intermixed. 
\par
The functional distribution of \textit{Tremblaya phenacola} PPER genes is not random either. 
The transcriptional machinery and the ribosomes are of betaproteobacterial origin, while aminoacyl-tRNA synthetases (not the complete set, as in other mealybugs) appear to be of gammaproteobacterial origin. 
The only exception is serS (a pseudogene in several \textit{Tremblaya princeps} strains; Husnik and McCutcheon, 2016), which gave no clear affiliation. 
Except for iscSUA (involved in (Fe–S) cluster assembly), genes devoted to tRNA maturation are also of gammaproteobacterial origin. 
This pattern is similar to the nested endosymbiotic consortia from pseudococcinae mealybugs, \textit{Tremblaya} has retained most of its own transcriptional and translational machinery except for aminoacyl-tRNA synthetases, which must be provided by the gammaendosymbiont. 
Furthermore, all maintained subunits of the DNA polymerase (also preserved in other \textit{Tremblaya princeps}) are of beta origin. 
However, the other proteins involved in DNA replication (helicase and ligase) are of gamma origin; the first one has been preserved in all other \textit{Tremblaya} genomes sequenced, while the second is absent in all of them. 
Genes involved in translation initiation (\textit{infA}, \textit{infB} and \textit{infC}) and elongation (\textit{fusA} and \textit{tufA}) are of beta origin, although there is an additional gammaproteobacterial \textit{infA}. 
Genes involved in translation termination (\textit{prfA}, \textit{prfB} and \textit{prmC}), ribosome recycling (\textit{frr}) and degradation of proteins stalled during translation (\textit{smpB}), as well as N-formyltransferase (\textit{fmt}) and peptide deformylase (\textit{def}) are of gamma origin.
\par
Like all other mealybug endosymbionts, \textit{Tremblaya phenacola} PPER mediates essential amino acid synthesis. 
As in most studied pseudococcinae mealybugs, all genes retained for the biosynthesis of methionine, threonine, isoleucine, leucine and valine, and the production of phenylalanine from chorismate are of betaproteobacterial origin, while the pathways for the production of chorismate and lysine retain the same patchwork pattern. 
Histidine biosynthesis is an exception, as PPER has only retained genes of gammaproteobacterial origin. 
The cysteine biosynthetic pathway is more complete in PPER, with all genes of gamma origin. 
Regarding tryptophan biosynthesis, dominated by gammaproteobacterial genes in previously analyzed mealybugs’ endosymbiotic consortia, in \textit{Tremblaya phenacola} PPER the first step is performed by beta proteins, while the rest of the genes are of gamma origin, a similar pattern to that found in other insect’s endosymbiotic consortia (that is, \textit{Serratia/Buchnera} in lachninae aphids and some \textit{Carsonella}/secondary systems in psyllids; Lamelas \textit{et al.}, 2011; Sloan and Moran, 2012; Manzano-Marín \textit{et al.}, 2016). 
\par
\textit{Tremblaya phenacola} PPER genome goes beyond what could be considered a standard horizontal gene transfer event, and rather resembles the complete fusion of two genomes to form a new chimeric organism. 
Independent phylogenomic analyses of two concatenations of the \textit{Tremblaya phenacola} PPER genes assigned as beta or gammaproteobacterial placed beta origin genes in \textit{Tremblaya phenacola} clade, while the gammaproteobacterial genes were placed into the \textit{Sodalis}-allied clade (Husnik and McCutcheon, 2016) as a sister species of ‘\textit{Candidatus} Mikella endobia’, nested gamma-endosymbiont of the pseudoccinae mealybug \textit{Paracoccus marginatus}. 
\par
How could the genomic fusion have ocurred? 
Although HGT is uncommon in modern endosymbionts, it is an extended phenomenon in flowering-plant mitochondria (Sanchez-Puerta, 2014), derived from an ancestral α-proteobacterial endosymbiont
(Andersson \textit{et al.}, 1998). 
The most notable case corresponds to \textit{Amborella trichopoda}, whose mitochondrial DNA has incorporated the complete mitochondrial genomes of three green algae and one moss, plus two mitochondrial genome equivalents from other angiosperms (Rice \textit{et al.}, 2013). 
Such a high frequency of HGT has been explained by mitochondrial fusion and subsequent genomes fusion and rearrangements, mediated by homologous recombination systems (Maréchal and Brisson, 2010). 
Something similar might have occurred in \textit{Tremblaya phenacola} PPER. 
On the basis of current evidences, the ancestor of all \textit{Tremblaya} probably had a reduced genome (Husnik and McCutcheon, 2016). 
In the lineage driving to \textit{Tremblaya phenacola} PPER, a gammaproteobacterium must have entered the consortium and, instead of replacing \textit{Tremblaya phenacola} (as in the tribe Rhizoecini and genus \textit{Rastrococcus}; Gruwell \textit{et al.}, 2010), or establishing a nested endosymbiosis (as in the \textit{Tremblaya phenacola} clade; reviewed by Husnik and McCutcheon, 2016), a cellular fusion event must have occurred, followed by genomic fusion. 
It cannot be discarded that a nested endosymbiosis preceded the cellular and genomic fusions. 
Because this phenomenon implies the existence of a DNA recombination machinery, the most plausible hypothesis is that such genes were present in the genome of the gammaproteobacterial donor, similarly to what has been described in citrus mealybug (López-Madrigal \textit{et al.}, 2013). 
In fact, most mealybugs’ gamma-endosymbionts that have been completely sequenced (McCutcheon and von Dohlen, 2011; López-Madrigal \textit{et al.}, 2013; Husnik and McCutcheon, 2016) or screened for homologous recombination genes (López-Madrigal \textit{et al.}, 2015) present a more or less complete recombination machinery. 
Transposable elements might also facilitate a fusion process. 
Some authors suggest that in arthropod intracellular environments, the possibility of two bacteria co-infecting the same cell generates an ‘intracellular arena’ where distantly related bacterial lineages can exchange mobile elements (Duron, 2013). 
However, although insertion sequences are frequent in early endosymbiotic stages (Latorre and Manzano-Marín, 2016), they have not been identified in any sequenced mealybugs’ gamma-endosymbiont, and no indication of their former presence in \textit{Tremblaya phenacola} PPER. 
After the fusion, the chimeric genome must have undergone massive gene loss, getting rid of almost all redundant and non-essential genes. 
The initial presence of homologs might have accelerated gene losses through recombination until DNA recombination genes disappeared. 
The remnant repeats involved in intrachromosomal recombination might have been maintained due to the loss of
such genes, leading to the current, complex genome organization.
\par
\textbf{Szabo \textit{et al.}, 2017, Convergent patterns in the evolution of mealybug symbioses involving different intrabacterial symbiosis.} \newline
Manna mealybug \textit{Trabutina mannipara} contains a betaproteobacterial symbiont \textit{Tremblaya}, which contains a gammaproteobacterial symbiont \textit{Trabutinella endobia}. 
Genomic sequences of \textit{Tremblaya} are highly syntenic and harbored nearly identical sets of genes in the manna/citrus mealybug in accordance with a monophyletic origin of the outer symbionts among mealybugs. 
The genome of the intrabacterial symbiont \textit{Trabutinella} shows only minimal synteny with that of \textit{Moranella} from citrus mealybug, which is consistent with the distinct evolutionary origin of these symbionts. 
Although \textit{Trabutinella} genome is much smaller than \textit{Moranella} genome (McCutcheon & von Dohlen, 2011), \textit{Trabutinella} genome is likely still in the process of reduction as indicated by the presence of 27 pseudogenes (Moran and
Bennett, 2014). 
\par
Genes of bacterial origin are found in manna mealybug genome. 
These genes are involved in synthesis of essential amino-acid, biotin and riboflavin. 
Many laterally acquired genes in genome of manna/citrus mealybug appear as sisters in phylogenetic tree, indicating that they share a common origin and were present in ancestral mealybugs before diversification of manna/citrus mealybug.
\par
Symbiotic partners of the manna mealybug and citrus mealybug systems partition the synthesis of essential amino acids in a highly similar manner. 
In most of the essential amino acid production pathways, exactly the same steps are carried out by the inner or the outer symbiont in manna mealybug and citrus mealybug, despite the independent origin of the intrabacterial symbionts. 
A similar, yet far less complex situation has been observed among members of Auchenorrhyncha, where \textit{Sulcia} synthesizes eight or seven essential amino acids while the remaining two or three are produced by different co-symbionts in different lineages, for instance. by \textit{Baumannia} in sharpshooters, \textit{Hodgkinia} in cicadas and \textit{Zinderia} in spittlebugs (McCutcheon and Moran, 2010; Bennett and Moran, 2013).
\par
A conceivable scenario explaining the observed similarities between the manna mealybug and citrus mealybug symbioses would be that before manna and citrus mealybug diverged, the \textit{Tremblaya} ancestor was already infected by an (intra-)bacterial symbiont in ancestral mealybugs and this ancient association would have facilitated reduction of the \textit{Tremblaya} genome and has shaped its gene repertoire. 
The inner symbiont might have been subsequently replaced in the ancestor of citrus and/or manna mealybug, with the new symbiont taking over the functions required by \textit{Tremblaya} and at the same time allowing for loss of further genes from the \textit{Tremblaya} genome, which would account for the observed differences between the two systems.
This scenario is favoured over alternative scenarios such as ancient associations of \textit{Tremblaya} with another bacteriocyte-associated symbiont because of the high level of congruence between the loss of genes in different \textit{Tremblaya} strains at intermediate steps of essential amino-acid synthesis pathways. 
\par
\textbf{Bublitz \textit{et al.}, 2019, Peptidoglycan production by an insect-bacterial mosaic citrus mealybug, Tremblaya, Mornaella.} \newline
Citrus mealybug \textit{Planococcus citri} has two bacterial symbionts: \textit{Tremblaya princeps} that lives in bacteriocytes of mealybug, and \textit{Moranella endobia} that lives in \textit{Tremblaya}. 
\par
Genomics predicts a complete pathway for synthesis of peptidoglycan, composed of \textit{Moranella} genes and several genes that were horizontally transferred from bacteria to the nuclear genome of mealybugs. 
No related genes are found in \textit{Tremblaya}. 
Peptidoglycan constitutes are detected in whole insect preparation, and the peptidoglycan-specific molecule D-Ala ss specifically localized at the \textit{Moranella} periphery. 
A peptidoglycan-targeting antibiotic specifically affects the \textit{Moranella} cell envelope. 
A peptidoglycan-related horizontal gene transfer of Alphaproteobacterial origin is localized to the \textit{Moranella} cytoplasm.
\par
Peptidoglycan-based cell wall is an ancient and defining feature of bacteria, and peptidoglycan biosynthesis pathway is often highly conserved in bacterial genomes. 
There are two exceptions of this pattern. 
The first example of peptidoglycan-related horizontal gene transfers comes from the chromatophore of the rhizarian protist \textit{Paulinella chromatophora}. 
\textit{Paulinella} chromatophore has a peptidoglycan layer (Kies, 1974), which is encoded primarily on the chromatophore genome with the exception of one bacterial horizontal gene transfer to the host protist genome (Nowack \textit{et al.}, 2016). 
The second example comes from the group of photosynthetic eukaryotes whose ancestor formed the original endosymbiosis with the cyanobacterium that became the chloroplast. 
This group, called the Archaeplastida, includes land plants, red algae, green algae, and glaucophyte algae (Lane and Archibald, 2008; McFadden, 2001). 
Many archaeplastidal nuclear genomes encode some peptidoglycan-related endosymbiosis gene transfers and horizontal gene transfers (van Baren \textit{et al.}, 2016; Sato and Takano, 2017), but these genes do not always seem to work together to form a functional peptidoglycan layer at the chloroplast periphery. 
A chloroplast-localized peptidoglycan layer has been verified using fluorescently labeled D-Ala in the moss \textit{Physcomitrella patens} (Hirano \textit{et al.}, 2016), and possible chloroplast peptidoglycan layers have been observed by EM in glaucophytes (Schenk, 1970). 
But in the land plant \textit{Arabidopsis thaliana}, which retains some peptidoglycan-related genes on its nuclear genome, no peptidoglycan layer exists at the chloroplast periphery and at least one peptidoglycan-related enzyme has been coopted for a different function (Garcia \textit{et al.}, 2008). 
These results serve as a cautionary note about inferring function from genomics alone: gene presence is not a reliable predictor of biological function (Doolittle, 2013).
\par
One important remaining question is the source of D-Ala and D-Glu in \textit{Moranella}’s peptidoglycan, as homologs of Alr (alanine racemase) and MurI (glutamate racemase) do not present as horizontal gene transfers on mealybug genome or present in \textit{Moranella} genome. 
These activities might be moonlighted by other genes. 
For example, GlyA and MetC have been shown to moonlight as alanine racemases in \textit{Chlamydia trachomatis} and \textit{Escherichia coli}, respectively (De Benedetti \textit{et al.}, 2014; Kang \textit{et al.}, 2011; Otten \textit{et al.}, 2018), and eukaryotic homologs for these genes present on the mealybug genome. 
Similarly, DapF has been shown to moonlight as a glutamate racemase in \textit{Chlamydia trachomatis} (Liechti \textit{et al.}, 2018), and this gene exists as an horizontal gene transfer of alphaproteobacterial origin on citrus mealybug genome (Husnik \textit{et al.}, 2013). 
It is also possible that the source of D-Ala and D-Glu is not from these putatively moonlighting enzymes at all, but rather from either the plant sap diet of the insect or from D-amino acids in citrus mealybug produced from normal insect
biochemistry. 
D-amino acids have been found in both plants (Robinson, 1976) and insects (Auclair and Patton, 1950; Corrigan and Srinivasan, 1966; Corrigan,1969), although the levels of these compounds have not been measured in citrus mealybug. 
\par
MurF encoded by horizontal transferred genes in mealybug genome is localized specifically in \textit{Moranella} cytoplasm. 
Importing enzyme/mRNA into \textit{Moranella} cytoplasm for peptidoglycan synthesis may help avoid immune responses. 
Other insects with long-term endosymbionts devote resources to scavenging peptidoglycan fragments in order to prevent continuous immune activation (Maire \textit{et al.}, 2019). 
By sequestering peptidoglycan production to inside of \textit{Moranella}, citrus mealybug may avoid the need for such contingency pathways, at least until \textit{Moranella} cells are recycled near the end of the mealybug’s life (Kono \textit{et al.}, 2008).
\par
Most peptidoglycan-related genes on mealybug genome (horizontal transferred) function in the cytoplasmic part of peptidoglycan synthesis, whereas the peptidoglycan-related genes retained by \textit{Moranella} all code for inner membrane- or periplasm-associated proteins. 
It indicates that genes functioning in \textit{Moranella} cytoplasm are more likely to be transferred to insect genome, and this may reflect the mechanisms through which proteins/RNA encoded by insect are transported into symbionts. 
\par
Host takeover of endosymbiont peptidoglycan production can be an important step in the regulation of endosymbiont cell division and potentially further integration with the host organism (de Vries and Gould, 2018). 
In moss, knocking out a peptidoglycan-related horizontal gene transfer on the nuclear genome results in enlarged chloroplasts
(Machida \textit{et al.}, 2006), and treatment with various peptidoglycan-targeting antibiotics results in fewer and larger chloroplasts per host cell (Katayama \textit{et al.}, 2003). 
Together these data suggest that the movement of peptidoglycan-related genes from organelle genome to the host is a way for hosts to regulate organelle division (de Vries and Gould, 2018; Katayama \textit{et al.}, 2003; Machida \textit{et al.}, 2006). 
In citrus mealybugs, \textit{Tremblaya} was acquired before \textit{Moranella} (Hardy \textit{et al.}, 2008; Thao \textit{et al.}, 2002), and so the host insect must have found a way of controlling \textit{Tremblaya} as the sole endosymbiont prior to the acquisition of \textit{Moranella}. 
It is tempting to speculate that peptidoglycan-related horizontal gene transfers have been retained on the insect genome as a way of controlling the cell division of a bacterium that lives inside of another bacterium inside of insect cells.
\par
\textbf{Husink \textit{et al.}, Horizontal gene transfer from diverse bacteria to an insect genome enables a tripartite nested mealybug symbiosis.} \newline
The smallest reported bacterial genome belongs to \textit{Tremblaya princeps}, a symbiont of \textit{Planococcus citri} mealybugs (PCIT). 
\textit{Tremblaya} PCIT not only has a 139 kb genome, but possesses its own bacterial endosymbiont, \textit{Moranella endobia}. 
Genome and transcriptome sequencing, including genome sequencing from \textit{Tremblaya} symbiont of \textit{Phenacoccus avenae} (PAVE), which lacks intracellular bacteria, reveals that the extreme genomic degeneracy of \textit{Tremblaya} PCIT likely resulted from acquiring \textit{Moranella} as an endosymbiont. 
In addition, at least 22 expressed horizontally transferred genes from multiple diverse bacteria to the mealybug genome likely complement missing symbiont genes. 
However, none of these horizontally transferred genes are from \textit{Tremblaya}, showing that genome reduction in this symbiont has not been enabled by gene transfer to the host nucleus.
\par
Acquisition of \textit{Moranella} symbiont may trigger extreme genome degeneracy in \textit{Tremblaya} PCIT. 
Genome reduction in \textit{Tremblaya} PAVE occurs to a degree consistent with other previously reported tiny symbiont genomes, and \textit{Tremblaya} PCIT gene set is an almost perfect subset of \textit{Tremblaya} PAVE. 
These results suggest that much of the reductive genome evolution observed in \textit{Tremblaya} (down to approximately 170 kb) occurred before the acquisition of \textit{Moranella} in the common ancestor of \textit{Planococcus citri} and \textit{Phenacoccus avenae} and that the extreme genomic degeneracy observed in \textit{Tremblaya} PCIT (from 170 kb to 140 kb) was likely due to the acquisition of \textit{Moranella} by \textit{Tremblaya} at some point in the lineage leading to \textit{Planococcus citri}. 
This scenario is consistent with studies showing that massive and rapid gene loss can occur in bacteria that transition to a symbiotic lifestyle (Mira \textit{et al.}, 2001; Moran & Mira, 2001; Nilsson \textit{et al.}, 2005), after which gene loss slows, and gross genomic changes become infrequent, even over hundreds of millions of years (McCutcheon and Moran, 2010; Tamas
\textit{et al.}, 2002; van Ham \textit{et al.}, 2003). 
\par
Pathways for translation, synthesis of essential amino acids, vitamins and peptidoglycan in \textit{Tremblaya} PCIT are complemented by \textit{Moranella}, mealybug genes originated from bacteria-to-mealybug horizontal gene transfers (HGTs) and mealybug genes of eukaryotic origin. 
In PCIT, ten HTGs group closely with other alphaproteobacterial sequences in phylogenetic trees, and nine HTGs from Gammaproteobacteria, two from Bacteroidetes, and one that is phylogenetically unresolved. 
The majority of these HGTs are not present in \textit{Tremblaya} and \textit{Moranella} genomes. 
\par
The presence of a large number of HTGs involved in peptidoglycan production and recycling is consistent with the hypothesis that cell lysis is the mechanism used to share gene products between \textit{Moranella} and \textit{Tremblaya} PCIT (Koga \textit{et al.}, 2013; McCutcheon and von Dohlen, 2011). 
This idea was initially suggested based on a lack of transporters encoded on the \textit{Moranella} genome combined with the large number of gene products or metabolites involved in essential amino acid biosynthesis and translation that would need to pass between \textit{Moranella} and \textit{Tremblaya} PCIT for the symbiosis to function (McCutcheon and von Dohlen, 2011). 
Subsequent electron microscopy on mealybugs closely related to PCIT showed that although most gammaproteobacterial cells infecting the \textit{Tremblaya} cytoplasm were rod shaped, some were amorphous blobs seemingly in a state of degeneration (Koga \textit{et al.}, 2013).
THe results suggest a plausible mechanism for how the insect host controls this process: by differentially controlling the
expression of the horizontally transferred genes, the host could regulate the cell wall stability of \textit{Moranella}. Increasing the expression of \textit{murABCDE} genes would increase the integrity of \textit{Moranella}’s cell wall, and increasing the expression of \textit{mltD}/\textit{amiD} would tend to decrease \textit{Moranella}’s cell wall strength. 
As \textit{Tremblaya} PCIT encodes no cell-envelope-related genes and likely uses host-derived membranes to define its cytoplasm, it would be unaffected by changes in gene expression related to peptidoglycan biosynthesis. 
This hypothesis is testable, because the levels of \textit{Tremblaya} and \textit{Moranella} are uncoupled in mealybugs closely related to PCIT; in males in particular, \textit{Moranella} levels drop to undetectable levels while \textit{Tremblaya} persists (Kono \textit{et al.}, 2008). \
In situations where \textit{Moranella} is reduced with respect to \textit{Tremblaya}, low expression of \textit{murABCDEF} and increased expression of \textit{mltD}/\textit{amiD} would be expected. 
\par
\textbf{McCutcheon and Dohlen, 2011, An interdependent metabolic patchwork in the nested symbiosis of mealybug.} \newline
Citrus mealybug \textit{Planococcus citri} represents a nested symbiosis system: 
a betaproteobacteria \textit{Candidatus} Tremblaya princeps lives inside citrus mealybug, while a gammaproteobacteria \textit{Candidatus} Moranella endobia lives in cytoplasm of \textit{Tremblaya}. 
\par
\textit{Tremblaya} genome is extremely small (0.14 Mbp) and degenerated (121 proteins). 
A 7-kbp region of \textit{Tremblaya} genome exists two orientations within single insect host, while \textit{Tremblaya} lacks genes involved in recombination. 
\par
\textit{Tremblaya} retains genes involved in essential amino acid synthesis, but does not have complete pathways of its own. \textit{Moranella} complements several essential amino acid synthesis genes lost in \textit{Tremblaya}. 
However, it is unclear how transport of metabolites occurs between cosymbionts. 
\textit{Tremblaya} genome encodes no predicted transporters. 
\textit{Moranella} encodes a handful of proteins involved in membrane transportation, but none are specific for amino acids
or their precursors. 
Some components of the Sec translocation machinery are present in the \textit{Moranella} genome, and it is possible that these are used to transport some proteins across \textit{Moranella}’s inner membrane. 
A search for signal peptides in the \textit{Moranella} proteome revealed 27 proteins with N-terminal secretory signal peptides; however, none was involved in essential amino acid biosynthesis. 
\par
Tremblaya is missing several gene homologs for translation-related functions that are often retained in other highly reduced bacteria genomes, including all aminoacyl-tRNA synthetases, translational release factors. 
As translation machinery is significantly different in eukaryotes and bacteria, it seems unlikely that the missing translation-related genes in \textit{Tremblaya} are complemented by host. 
Horizontal gene transfer from bacteria to host might be the solution, although no transfer of functional genes
between symbiont and host has been found in another two insects, pea aphid (Nikoh \textit{et al.}, 2010) and human body louse (Kirkness \textit{et al.}, 2010). 
\par
The nested structure of the mealybug symbionts is likely controlled by the host. 
There are at least two morphological forms of \textit{Moranella}: a reproductive form in which cells were small in size and in the process of dividing, and a degenerative phase in which cells became unevenly shaped and elongated (Buchner, 1965). 
The particular Moranella form was dependent on the life stage of the insect and seemed to be synchronized within a bacteriocyte (Buchner, 1965). 
Furthermore, the infection levels of Tremblaya and Moranella are uncoupled in mealybugs (Kono \textit{et al.}, 2008). 
During male development, the number of Moranella cells relative to \textit{Tremblaya} cells drops significantly as the insects age, whereas in female insects, the levels of the two symbionts remain roughly equivalent over the entire life cycle (Kono \textit{et al.}, 2008). 
Given that Tremblaya has an extremely limited coding capacity that is largely devoted to essential amino acid biosynthesis and translation, and given that only seven genes are of completely unknown function, it seems impossible that \textit{Tremblaya} itself controls any structural aspect of the symbiosis. 
Likewise, the \textit{Moranella} genome does not encode any genes involved in traditional infective strategies and does not indicate any obvious pathway by which it could be an active participant involved in seeking out the \textit{Tremblaya} cytoplasm. 
Thus, it seems likely that the host is largely in control of the structure and organization of this bacteria-within-a-bacterium symbiosis.
\par
\textit{Tremblaya} survives with highly reduced genome with loss of genes thought to be essential for survival (\textit{e.g.} translation). 
The missing activities can be complemented by several mechanisms: 
(1) gene products or metabolites of either host or bacterial origin imported from the host; 
(2) gene products or metabolites imported directly from the other symbionts if present; 
(3) genetic coadaptations to the loss of genes within the reduced genome itself; 
(4) the direct use of \textit{Moranella} gene products as a result of a simple, passive mechanism such as \textit{Moranella} cell lysis within the cell membrane system of \textit{Tremblaya}. 
\par
\textit{Tremblaya} genome is extremely small, but low gene dense. 
During the shift from a free-living to an obligate intracellular lifestyle, where the constant exposure to the stable and rich environment of the host cell combined with a severe reduction in population size (and subsequent reduction in the efficacy of purifying selection) allows large numbers of pseudogenes to accumulate (Ochman \textit{et al.}, 2006; Andersson \textit{et al.}, 2001). 
These pseudogenes are eventually purged from the genome through mutational patterns favoring deletions (Mira \textit{et al.}, 2001), leading to small gene-dense genomes such as those from insect nutritional symbionts. 
A possible explanation is that \textit{Tremblaya} undergone genome reduction after association with mealybug, and acquisition of \textit{Moranella} leads to further genome reduction. 
Basal lineages of mealybugs in the same subfamily as citrus mealybug seem to contain \textit{Tremblaya} without the intracellular gammaproteobacterial endosymbiont (Hardy \textit{et al.}, 2008; Thao \textit{et al.}, 2002), indicating that \textit{Moranella} was acquired after the establishment of \textit{Tremblaya}. 
The patterns of gene pseudogenization also fit this hypothesis, as most pseudogenized \textit{Tremblaya} genes have functional \textit{Moranella} homologs.
\par
\textbf{Gomez-Polo \textit{et al.}, 2017, An exceptional family: Ophiocordyceps-allied fungus dominates the microbiome of soft scale insects (Hemiptera Sternorrhyncha: Coccidae).} \newline
Ribosomal genes from seven soft scale (Coccidae) species showed high prevalence of an \textit{Ophiocordyceps}-allied fungal symbiont, which is from an lineage widely known as entomopathogenic. 
The \textit{Ophiocordyceps}-allied fungus from soft scales is closely related to fungi described from other hemipterans, and they appear to be monophyletic, although the phylogenies of the \textit{Ophiocordyceps}-allied fungi and their hosts do not appear to be congruent. 
Microscopic observations show that the fungal cells are lemon-shaped, are distributed throughout the host’s body and are
present in the eggs, suggesting vertical transmission.
\par
\textbf{Deng \textit{et al.}, 2021, The ubiquity and development-related abundance dynamics of Ophiocordyceps fungi in soft scale insects.} \newline
Nuclear ribosomal internal transcribed spacer (ITS) gene fragment was used to analyze the diversity of fungal communities in 28 soft scale (Coccidae) species. 
Coccidae-associated \textit{Ophiocordyceps} fungi (COF) were prevalent in all 28 tested species with high relative abundance. 
Meanwhile, the first and second instars of \textit{C. japonicus} had high relative abundance of COF, while the relative
abundances in other stages were low, ranging from 0.68\% to 2.07\%. 
The result of fluorescent in situ hybridization showed that the COF were widely present in \textbf{hemolymph} and vertically transmitted from mother to offspring. 
\par
\textbf{Chong & Moran, 2016, Intraspecific genetic variation in hosts affects regulation of obligate heritable symbionts.} \newline
The extent of intraspecific variation in the regulation of a mutually obligate symbiosis, between the pea aphid \textit{Acyrthosiphon pisum} and its maternally transmitted symbiont \textit{Buchnera aphidicola}, using experimental crosses to identify effects of host genotypes. 
Symbiont titer, as the ratio of genomic copy numbers of symbiont and host, as well as developmental time and fecundity of hosts, were measured. 
There was a large (>10-fold) range in symbiont titer among genetically distinct aphid lines harboring the same \textit{Buchnera} haplotype. 
Aphid clones also vary in fitness, measured as developmental time and fecundity, and genetically based variation in titer is correlated with host fitness, with higher titers corresponding to lower reproductive rates of hosts. 
The results show that obligate symbiosis is not static but instead is subject to short-term evolutionary dynamics, potentially reflecting coevolutionary interactions between host and symbiont.
\par
\textbf{Henry \textit{et al.}, 2013, Horizontally Transmitted Symbionts and Host Colonization of Ecological Niches.} \newline
Facultative or “secondary” symbionts are common in eukaryotes, particularly insects. 
While not essential for host survival, they often provide significant fitness benefits. 
It has been hypothesized that secondary symbionts form a “horizontal gene pool” shuttling adaptive genes among host lineages in an analogous manner to plasmids and other mobile genetic elements in bacteria. 
However, we do not know whether the distributions of symbionts across host populations reflect 
random acquisitions followed by vertical inheritance or whether 
the associations have occurred repeatedly in a manner consistent with a dynamic horizontal gene pool. 
Here we explore these questions using the phylogenetic and ecological distributions of secondary symbionts carried by 1,104 pea aphids, \textit{Acyrthosiphon pisum}. 
We find that not only is horizontal transfer common, but it is also associated with aphid lineages colonizing new ecological niches, including novel plant species and climatic regions. 
Moreover, aphids that share the same ecologies worldwide have independently acquired related symbiont genotypes, suggesting significant involvement of symbionts in their host’s adaptation to different niches. 
We conclude that the secondary symbiont community forms a horizontal gene pool that influences the adaptation and distribution of their insect hosts. 
These findings highlight the importance of symbiotic microorganisms in the radiation of eukaryotes.
\par
\textbf{Hosokawa \textit{et al.}, 2007, Obigate symbiont involved in pest status of host insect.} \newline
A pest stinkbug species, \textit{Megacopta punctatissima}, performed well on crop legumes, while a closely related non-pest species, \textit{Megacopta cribraria}, suffered low egg hatch rate on the plants. 
When their obligate gut symbiotic bacteria were experimentally exchanged between the species, their performance on the crop legumes was completely reversed: 
the pest species suffered low egg hatch rate, whereas the non-pest species restored normal egg hatch rate and showed good performance. 
The low egg hatch rates were attributed to nymphal mortality before or upon hatching, which were associated with the symbiont from the non-pest stinkbug irrespective of the host insect species.
\par
\textbf{Couret \textit{et al.}, 2019, Even obligate symbioses show signs of ecological contingency: Impacts of symbiosis for an invasive stinkbug are mediated by host plant context.} \newline
Many species interactions are dependent on environmental context, yet the benefits of obligate, mutualistic microbial symbioses to their hosts are typically assumed to be universal across environments. 
We directly tested this assumption, focusing on the symbiosis between the sap-feeding insect \textit{Megacopta cribraria} and its primary bacterial symbiont \textit{Candidatus} Ishikawaella capsulata. 
We assessed host development time, survival, and body size in the presence and absence of the symbiont on two alternative host plants and in the insects' new invasive range. 
We found that association with the symbiont was critical for host survival to adulthood when reared on either host plant, with few individuals surviving in the absence of symbiosis. 
Developmental differences between hosts with and without microbial symbionts, however, were mediated by the host plants on which the insects were reared. 
Our results support the hypothesis that benefits associated with this host–microbe interaction are environmentally contingent, though given that few individuals survive to adulthood without their symbionts, this may have minimal impact on ecological dynamics and current evolutionary trajectories of these partners.
\par
\textbf{Arp \textit{et al.}, 2016, Annotation of the Asian citrus psyllid genome reveals a reduced innate immune system.}
A genome-wide analysis of immune genes in Asian citrus psyllids \textit{Diaphorina citri} was presented. 
$\beta$-1,3-glucan binding protein and IMD signaling (Imd, Dredd and Relish) are absent. 
Antimicrobial peptides are absent. 
\par
\textbf{Sloan \textit{et al.}, 2014, Parallel histories of horizontal gene transfer facilitated extreme reduction of endosymbiont genome in sap-feeding insects.} \newline
Endosymbionts typically have experienced extreme genome reduction, but the role of host in this process remains unclear. 
\textit{Carsonella ruddii} is a vertically-transmitted gammaproteobacterial endosymbiont widely-present in psyllid bacteriomes. 
It lacks genes involved in DNA replication, transcription and translation. 
There are three possible mechanisms for the exceptional gene loss: 
(1) modification in cellular processes or selection for multifunctional proteins; 
(2) compensation from additional endosymbionts, as observed in psyllid \textit{Ctenarytaina eucalypti}; 
(3) compensation from host-coding proteins. 
mRNA-seq data of psyllid \textit{Pachypsylla venusta} revealed that host genes that were up-regulated in bacteriome complement amino acid synthesis pathways that are absent/incomplete in endosymbionts. 
Draft genome of host revealed horizontal gene transferrs (HGTs) from bacteria of diverse lineage. 
\par
\textbf{Salcedo-Porras \textit{et al.}, 2019, \textit{Rhodnius prolixus}: identification of missing components of the IMD immune signaling pathway and functional characterization of its role in eliminating bacteria.}
Previously reported missing components of IMD pathway were found in \textit{Rhodnius prolixus}. 
They were involved in response to infection with Gram-negative bacteria. 
RNAi revealed the role of IMD pathway in regulating antimicrobial peptides (AMPs). 
\par
\textbf{Bennett \textit{et al.}, 2014, Differential Genome Evolution Between Companion Symbionts in an Insect-Bacterial Symbiosis.} \newline
To understand how symbiont genome degeneration proceeds, we compared the genomes of symbionts in two leafhopper species, \textit{Homalodisca vitripennis} (glassy-winged sharpshooter [GWSS]) and \textit{Graphocephala atropunctata} (blue-green sharpshooter [BGSS]) (Hemiptera: Cicadellidae). 
Each host species is associated with the anciently acquired Candidatus \textit{Sulcia muelleri} (Bacteroidetes) and the more recently acquired Candidatus \textit{Baumannia cicadellinicola} (Gammaproteobacteria). 
BGSS Ca. \textit{Baumannia} (B-BGSS) retains 89 genes that are absent from GWSS Ca. \textit{Baumannia} (B-GWSS); these underlie central cellular functions, including cell envelope biogenesis, cellular replication, and stress response. 
In contrast, Ca. \textit{Sulcia} strains (S-GWSS, S-BGSS) differ by only a few genes.
\par
Because Ca. \textit{Baumannia cicadellinicola} and Ca. \textit{Sulcia muelleri} in these hosts diverged in synchrony with their host lineage and with each other, the time of divergence for B-GWSS and B-BGSS is the same as that for S-GWSS and S-BGSS, implying that differences in pairwise divergences are due to differences in rates of nucleotide base substitution. 
\textit{Baumannia} strains have lower genome pairwise identity than \textit{Sulcia}. 
\textit{Baumannia} strains have high substitution rates (synonymous/non-synonymous), while substitution rates (synonymous/non-synonymous) of \textit{Sulcia} are often close to 0. 
Average dN/dS of \textit{Baumannia} and \textit{Sulcia} are below 1, indicating purifying selection. 
\par
\textbf{Ghosh \textit{et al.}, 2015, Prevalence and genetic diversity of endosymbiotic bacteria infecting cassava whiteflies in Africa.} \newline
The genetic diversity of field-collected whitefly from Tanzania, Malawi, Uganda and Nigeria was determined by mitochondrial DNA based phylogeny and restriction fragment length polymorphism. 
Cassava in these countries was infected with five whitefly populations, and each one was infected with different endosymbiotic bacteria. 
Incidences of Arsenophonus, Rickettsia, Wolbachia and Cardinium varied amongst the populations. 
Wolbachia was the most predominant symbiont with infection levels varying from 21 to 97\%. 
Infection levels of Arsenophonus varied from 17 to 64\% and that of Rickettsia was 0 to 53\%. 
Hamiltonella and Fritschea were absent in all the samples. 
Multiple locus sequence typing identified four different strains of Wolbachia infecting cassava whiteflies. 
A common strain of Wolbachia infected the whitefly population Sub-Saharan Africa 1-subgroup 1 (SSA1-SG1) and SSA1-SG2, while others were infected with different strains. 
Phylogeny based on 16S rDNA of Rickettsia and 23S rDNA of Arsenophonus also identified distinct strains.
\par
\textbf{Wernegreen & Moran, 1999, Evidence for genetic drift in endosymbionts (Buchnera): analyses of protein-coding genes.} \newline
Combined, these results suggest a decreased effectiveness of purifying selection in purging endosymbiont populations of slightly deleterious mutations, particularly those affecting codon usage and amino acid identity.
\par
\textbf{Kwak \textit{et al.}, 2022, Chromosomal-level assembly of \textit{Bactericera cockerelli} reveals rampant gene family expansions impacting genome structure, function and insect-microbe-plant-interactions.} \newline
\par
\textbf{Owen \textit{et al.}, 2020, Hemiptera phylogenomic resources: tree-based orthology prediction and conserved exon identification.} \newline
\par
\textbf{Ma \textit{et al.}, 2020, JNK pathway plays a key role in the immune system of the pea aphid and is regulated by microRNA-184.} \newline
\par
\textbf{Tomizawa \textit{et al.}, 2020, Numerous peptidoglycan recognition protein genes expressed in the bacteriome of the green rice leafhopper \textit{Nephotettix cincticeps} (Hemiptera, Cocadellidae).} \newline
\par
\textbf{Ma \textit{et al.}, 2021, Comparative analysis of \textit{Adelphocoris suturalis} Jakovlev (Hemiptera: Mirodae) immune responses to fungal and bacterial pathogens.} \newline
\par
\textbf{Yu \textit{et al.}, 2021, Characterization of PGRP-LB and immune deficiency in the white-backed planthopper \textit{Sogatella furcifera} (Hemiptera: Delphacidae).} \newline
\par
\textbf{Nishide \textit{et al.}, 2019, Functional crosstalk across IMD and Toll pathways: insight into the evolution of incomplete immune cascades.}
\end{sloppypar}
\end{document}