\documentclass[11pt]{article}

% \usepackage[UTF8]{ctex} % for Chinese 

\usepackage{setspace}
\usepackage[colorlinks,linkcolor=blue,anchorcolor=red,citecolor=black]{hyperref}
\usepackage{lineno}
\usepackage{booktabs}
\usepackage{graphicx}
\usepackage{float}
\usepackage{floatrow}
\usepackage{subfigure}
\usepackage{caption}
\usepackage{subcaption}
\usepackage{geometry}
\usepackage{multirow}
\usepackage{longtable}
\usepackage{lscape}
\usepackage{booktabs}
\usepackage{natbibspacing}
\usepackage[toc,page]{appendix}
\usepackage{makecell}
\usepackage{amsfonts}
 \usepackage{amsmath}

\usepackage[backend=bibtex,style=authoryear,sorting=nyt,maxnames=1]{biblatex}
\bibliography{} % Reference bib

\title{Abstracts: Hymenoptera}
\author{}
\date{}

\linespread{1.5}
\geometry{left=2cm,right=2cm,top=2cm,bottom=2cm}

\setlength\bibitemsep{0pt}

\begin{document}
\begin{sloppypar}
  \maketitle

  \linenumbers
\textbf{Evans \textit{et al.}, 2006, Immune pathways and defence mechanisms in honey bees \textit{Apis mellifera}} \newline
A genome-wide analysis of immune genes in honey bees is presented. 
In Toll pathway, transcription factor Dif is absent. 
Imd and JNK signaling is conserved. 
JAK/STAT pathway is conserved except extracellular cytokine \textit{Unpaired}. 
Comparing 17 immune gene families showed a decrease in paralog numbers in honeybees (71) compared with \textit{Drosophila melanogaster} (196) and \textit{Anopheles gambiae} (209). 
\par
Several hypotheses were proposed to explain the observation that honeybee immune gene families are systematically smaller than fruit flies and mosquitoes. 
First, there is a systematic downward bias in paralog counts in honeybee genome compared with fruit fly and mosquito. 
There is no evidence for such systematic decrease. 
Second, bee immune genes were missed due to sequence divergence. 
This should apply particularly for small peptides, \textit{e.g.} antimicrobial peptides. 
However, 1:1:1 orthologue presents in majority of analyzed immune gene families. 
Third, there are undescribed mechanisms for bees to broaden immune efficacy with reduced immune gene diversity. 
Forth, bees face low pressure from pathogens. 
Bee pathogens are diverse, but common ones are restricted. 
Fifth, social immunity in bees lessens pathogen pressure. 
Social immunity including reltentless hygiene, removing alien organisms and secreting antimicrobial substances, has been shown to be successful in honeybees. 
\par
\textbf{Barribeau \textit{et al.}, 2015, A depauperate immune repertoire precedes evolution of sociality in bees.} \newline
We find that the immune systems of these bumblebees, two species of honeybee, and a solitary leafcutting bee, are strikingly similar. 
Transcriptional assays confirm the expression of many of these genes in an immunological context and more strongly in young queens than males, affirming Batemanâ€™s principle of greater investment in female immunity. 
We find evidence of positive selection in genes encoding antiviral responses, components of the Toll and JAK/STAT pathways, and serine protease inhibitors in both social and solitary bees. 
Finally, we detect many genes across pathways that differ in selection between bumblebees and honeybees, or between the social and solitary clades.
\par
\textbf{Horak \textit{et al.}, 2020, Symbionts shape host innate immunity in honeybees.} \newline
\par
\textbf{Bull \textit{et al.}, 2012, A strong immune response in young adult honeybees masks their increased susceptibility to infection compared to older bees.}
\par
\textbf{Hou \textit{et al.}, 2021, Genome-Wide Analysis of Gene Families of Pattern Recognition Receptors in Fig Wasps (Hymenoptera, Chalcidoidea).} \newline
Based on the genomic data of 12 fig wasp species, with seven pollinators and five non-pollinators, we investigated the evolution patterns of pattern recognition receptors, such as Gram-negative bacteria-binding proteins (GNBPs), C-type lectins (CTLs), scavenger receptors class B (SCRBs), fibrinogen-related proteins (FREPs), galectins, and thioester-containing proteins (TEPs). 
Results showed that pollinators had no GNBP, but non-pollinators all had two gene members, which were clustered into two different clades in the phylogenetic tree, with each clade having specific domain and motif characteristics. 
The analysis of CTL and SCRB gene families also showed that there were lineage-specific genes and specific expansion in non-pollinators. 
Our results showed that there were significant differences in immune recognition between pollinators and non-pollinators, and we concluded that they had undergone flexible adaptive evolution in different environments.

\end{sloppypar}
\end{document}